Search suffix tree implementation

hannojg opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

This PR aims at improving the local search speed.
On a lower end android phone with a hightraffic account searching is **currently taking around 344ms** in the release build.
Our proposed **suffix tree search implementation takes 0.14ms** for searching on that same device.
**That is a 2456x improvement.**

<details>
<summary><b>Remaining todos:</b></summary>

- [x] @Szymon20000 Investigate size of `N`. Initially it was 1_000_000, I set it to be 25_000, because otherwise allocating the memory took a very long time. Can we approximate N? Can we use array buffers here?
- [x] @szymon20000 we need to support numbers as well. There are users who's id is for example a phone number, searching for that phone number is currently not possible
- [x] @Szymon20000 We need to support other unicode characters other than a-z (or need a way to handle it properly). For example Spanish or French letters. One example would be if we migrate the emoji trie to also use our suffix tree the emoji keywords in Spanish have special characters such as `ñ`
- [x] The faster recursive depth search function is currently broken
- [x] There seems to be a bug where while we are searching the tree currently gets recalculated, ie. the input data don't appear to be stable (could be due to react strict mode, but I think that shouldn't matter here). Would be good to investigate why that is (could be a separate ticket / PR) <- it happens because the first render gives personalDetails and recentReports as empty arrays, and then we recalculate memo again when we get actual data. useMemo is not called twice in StrictMode - it's applicable only to useEffect. Update during re-render happens because we send HTTP request which can update reports in ONYX and it'll lead to searchOptions recalculation
- [x] We need to cleanup the suffix tree implementation. Right now there is lots of code in the `ChatFinderPage` which we should hide in our implementation, such as the delimiter chars, the `stringToArray` function, the manual mapping from indexes to `OptionData`
- [x] Add all search values for personal details and reports into the search string for the tree 
- [x] Add the userToInvite case back, for now I skipped implementing this (I deleted the code, but I believe we need to keep the functionality)
- [x] Properly test and finish this PR
	- [x] Marc produces different search output (https://github.com/Expensify/App/commit/1d11ed2d534a13b3f17dbeb5ed836489ffef6c25)
	- [x] Taras can not be found
- [ ] Make a follow up issue to propose to [replace the emoji trie implementation with our suffix tree](https://expensify.slack.com/archives/C05LX9D6E07/p1725530302498229?thread_ts=1723209631.007999&cid=C05LX9D6E07) 
</details>

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/46591
PROPOSAL:


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- Open the chat finder page (search icon on the start page when authenticated)
- Perform searches, confirm that everything is found, make sure to also search for numbers
- Search for an email address (e.g. `hanno@margelo.io`), make sure it's found. Then search for `hannomargeloio` (removing all special chars), and make sure its still found
- You might want to open the production version and compare your search results with the dev version to make sure you got the same results  

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

Same as testing steps 

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

Same as testing steps 

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->


https://github.com/user-attachments/assets/57bca41b-6f56-4f5a-ad23-bf25ba9465d5



</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/2c62b027-6d00-42f3-825d-28460cdbd95f


</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->


https://github.com/user-attachments/assets/1102d13d-6bdf-41f8-b319-cff930399c71



</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->


https://github.com/user-attachments/assets/94f08c09-e0c2-479c-a64f-77cf40f5c8f9



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->


https://github.com/user-attachments/assets/fabbdaac-005d-4e74-8fec-72c8dac31e16



</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/597b8c09-9406-4aa4-90c9-18c9a7af7b00

</details>

hannojg committed:
suffix tree impl

hannojg committed:
add some helpful comments

hannojg committed:
example implementation usage of Suffixtree

hannojg commented:
cc @kirillzyusko

kirillzyusko committed:
fix: resolved one TODO

kirillzyusko committed:
fix: reduce code duplication

kirillzyusko committed:
refactor: O(2) -> O(1)

kirillzyusko committed:
refactor: minus one TODO

roryabraham commented on src/libs/SuffixUkkonenTree.ts:
```diff
@@ -0,0 +1,310 @@
+    const stringToSearch = `${strings.join('')}|`; // End Character
+    console.log('building search strings', performance.now() - start1);
+
+    const a = stringToArray(stringToSearch);
```
Just to get ahead of this, we'll definitely want descriptive names for some of these variables

kirillzyusko commented:
I'll copy TODOs and paste here, since I took this PR over but can't edit an original post, so will track my progress in this comment:

**Remaining todos:**

- [ ] @Szymon20000 Investigate size of N. Initially it was 1_000_000, I set it to be 25_000, because otherwise allocating the memory took a very long time. Can we approximate N? Can we use array buffers here?
- [ ]  @Szymon20000 we need to support numbers as well. There are users who's id is for example a phone number, searching for that phone number is currently not possible
- [ ]  @Szymon20000 We need to support other unicode characters other than a-z (or need a way to handle it properly). For example Spanish or French letters. One example would be if we migrate the emoji trie to also use our suffix tree the emoji keywords in Spanish have special characters such as ñ
- [ ]  We probably want to have an "update" function for the tree when for example the personal details are updated (instead of recreating the tree from scratch) - this needs to be added to the tree implementation as well @Szymon20000
- [ ]  There seems to be a bug where while we are searching the tree currently gets recalculated, ie. the input data don't appear to be stable (could be due to react strict mode, but I think that shouldn't matter here). Would be good to investigate why that is (could be a separate ticket / PR)
- [x]  We need to cleanup the suffix tree implementation. Right now there is lots of code in the ChatFinderPage which we should hide in our implementation, such as the delimiter chars, the stringToArray function, the manual mapping from indexes to OptionData
- [ ]  Add all search values for personal details and reports into the search string for the tree
- [ ]  Add the userToInvite case back, for now I skipped implementing this (I deleted the code, but I believe we need to keep the functionality) <- working on this
- [ ]  Properly test and finish this PR
- [ ]  Make a follow up issue to propose to [replace the emoji trie implementation with our suffix tree](https://expensify.slack.com/archives/C05LX9D6E07/p1725530302498229?thread_ts=1723209631.007999&cid=C05LX9D6E07)Remaining todos:

kirillzyusko committed:
fix: bring back userToInvite

kirillzyusko committed:
fix: make Marc discoverable again (when we search in second array, then we will always get  - so we add +1 bias)

kirillzyusko committed:
Merge branch 'main' into perf/search-suffix-ukkonen-tree

hannojg committed:
add debug timings

hannojg committed:
Merge branch 'main' of github.com:Expensify/App into perf/search-suffix-ukkonen-tree

hannojg committed:
adjust values for search

hannojg committed:
correct display name for personal details search

hannojg committed:
deduplicate search results

hannojg committed:
unify string cleaning

hannojg committed:
fix search function temporarily

hannojg committed:
make it work with numbers by base26 everything thats bigger than our alphabet

hannojg committed:
don't add delimiter char at the end

hannojg committed:
support for unicode

hannojg committed:
extended test case to make sure right item is returned

hannojg committed:
use concat instead of for-loop

hannojg committed:
Revert "use concat instead of for-loop"

This reverts commit 0806fb0f6ccf40f7fa4d958d92b5fef681fde0e6.

hannojg committed:
add note

hannojg committed:
wip: make t dynamic instead of preallocating

hannojg committed:
wip: refactor r list

hannojg committed:
wip: refactor l list

hannojg committed:
wip: refactor p list

hannojg committed:
wip: refactor s list

hannojg committed:
removed preallocations!

hannojg committed:
better naming

hannojg committed:
renamings

hannojg committed:
add userToToInvite to header count back

hannojg committed:
remove debug logs

hannojg committed:
cleanup

hannojg committed:
add unit test for special char handling

hannojg committed:
wip: document suffix tree functions

hannojg committed:
fix recursive search function

hannojg committed:
small cleanup

hannojg committed:
clean search input string

hannojg committed:
fixed flaky test

hannojg committed:
made test more reliable

hannojg committed:
migrate to useOnyx

hannojg committed:
ignore warning, this algorithm is optimized for speed

hannojg committed:
add timings to new search

hannojg committed:
refactor: separate suffix tree from search logic

hannojg committed:
add explanations + fix test

hannojg committed:
code documentation

hannojg committed:
add test verifying identical words work in suffix tree

hannojg committed:
add failing unit tests for duplicated string issue

hannojg committed:
add comment

hannojg committed:
fix search not returning results with same search value

hannojg committed:
perf: precalculate base26 map, optimize math operations, letter lookup table instead of charCodeAt, optimize push operations away

hannojg committed:
experiment: use array buffers

hannojg committed:
add correctChar optimization back

hannojg committed:
wip: optimizing search string construction

hannojg committed:
add unit test

hannojg committed:
Optimize performance by using one ArrayBuffer for search values and sharing that between operations

Squashed commit of the following:

commit fadc1d28e1ad565c8da03ff5168a09938c2e8e86
Author: Hanno J. Gödecke <die.drei99@yahoo.de>
Date:   Fri Sep 27 18:33:44 2024 +0200

    fix occurrence index values not fitting Int8 (numbers are larger than a byte ofc)

commit b582287b6f5d6930cf2e086f750e36a4cb92e3cf
Author: Hanno J. Gödecke <die.drei99@yahoo.de>
Date:   Fri Sep 27 15:15:00 2024 +0200

    fix tests

commit 523901b6539d174f277f90934fe333bc816eef92
Author: Hanno J. Gödecke <die.drei99@yahoo.de>
Date:   Fri Sep 27 15:14:04 2024 +0200

    optimize Int32Array -> Int8Array for search strings, as its all mapped to 0 - 26

commit 20f5c361a1fb7bc3a8d741c37ead68b500af562b
Author: Hanno J. Gödecke <die.drei99@yahoo.de>
Date:   Fri Sep 27 15:08:16 2024 +0200

    explain index better

commit 9dc9c0ed8c0040697cd939583d9b95dc85000c4f
Author: Hanno J. Gödecke <die.drei99@yahoo.de>
Date:   Fri Sep 27 15:05:05 2024 +0200

    add explaining code comments

commit d6f79d0b5ace740ddff70f2f2e727af15291fabc
Author: Hanno J. Gödecke <die.drei99@yahoo.de>
Date:   Fri Sep 27 15:00:49 2024 +0200

    search use new clamp option

commit 88ca64ccc66dc7f81d442a368e4e5bc8ec63100e
Author: Hanno J. Gödecke <die.drei99@yahoo.de>
Date:   Fri Sep 27 14:59:36 2024 +0200

    fix char set to skip not being respected

commit 5a8ba90d8531fbcb3911b66182c1227f35ab896a
Author: Hanno J. Gödecke <die.drei99@yahoo.de>
Date:   Fri Sep 27 14:41:39 2024 +0200

    remove console logs

commit 8d1b8b5afeaef6a33ced7536f5d224f69ce5904b
Author: Hanno J. Gödecke <die.drei99@yahoo.de>
Date:   Fri Sep 27 14:41:31 2024 +0200

    improve base26 conversions:

    - don't precalculate table, as we lookup too many chars that we never need
    - directly convert to numbers, not string representation, as we insert numbers (optimized 2 conversion steps)

commit 26acbe52cfcee6f4b535b8e79c2974e3046a621a
Author: Hanno J. Gödecke <die.drei99@yahoo.de>
Date:   Fri Sep 27 14:16:56 2024 +0200

    fix search broken for unicode chars

commit 8d63f452ecfd6a53fd8041c7b9bc75e5117c9a70
Author: Hanno J. Gödecke <die.drei99@yahoo.de>
Date:   Fri Sep 27 14:13:21 2024 +0200

    clean ups

commit 209da8d7204ec03ef444ac525990a634932d46c1
Author: Hanno J. Gödecke <die.drei99@yahoo.de>
Date:   Thu Sep 26 16:55:49 2024 +0200

    wip: fixing search

commit d1236c98244e704a7ed08ac186e66388d0d90f1f
Author: Hanno J. Gödecke <die.drei99@yahoo.de>
Date:   Thu Sep 26 16:03:54 2024 +0200

    wip: working, search broken

    Co-authored-by: Szymon Kapała <szymon20000@gmail.com>

hannojg committed:
ignore eslint warning; performance optimization

hannojg committed:
rename test file

hannojg commented:
Interesting, the reassure performance issues - indicating that the re-render count increased - is due to migrating to the `useOnyx` hook. Not sure how to handle, [asked here](https://margelo.slack.com/archives/C01GTK53T8Q/p1727461806388989).

hannojg committed:
add note

hannojg committed:
Update src/pages/ChatFinderPage/index.tsx

Co-authored-by: Gandalf <whogandalf@gmail.com>

hannojg committed:
wording

hannojg commented:
Okay, the PR is ready, and I am going to perform all review steps now. 

Whats missing in this PR is to save the creates "FastSearch" (tree) instance to a context, so it can be reused (so that we don't need to recreate the tree from scratch all the time).
I figures that will add another 200-400 lines of code and I wanted to make the lives of people reviewing not too hard - I will open an issue and do that as a follow up.

The performance of creating the "FastSearch" instance should on average still be faster than how we did it before, so it should be fine merging this one for now (but we'll get best performance when properly memoizing the tree).

BhuvaneshPatil commented:
@hannojg From my understanding, `withOnyx` waits till we get the data from onyx. Which is not case of `useOnyx`, and as we get data from `useOnyx` it re-renders the component.

We get loading status with `useOnyx`, we can use that. But I am not sure how we will reduce re-render count

roryabraham commented:
Helpful video on Ukkonen's algorithms and suffix trees: https://www.youtube.com/watch?v=OT5CigmVfh0

roryabraham reviewed (commented):
This is very impressive! It's definitely complex, but I think implementing Ukkonen's algorithm in a very readable way would be quite challenging (I've started attempting it in a separate branch, and I'm still a long way off 😅). Maybe we can make it more approachable with comments?

The most intuitive written explanation for this algorithm I've found (on which both the video and the visualization are based) is this SO: https://stackoverflow.com/questions/9452701/ukkonens-suffix-tree-algorithm-in-plain-english/9513423#9513423

If possible, I think it would be great to try and capture some of those high-level concepts and keywords in our implementation if they're applicable, so it has a better chance of being understood by future developers (i.e: `activePoint` as a triple of `(activeNode, activeEdge, activeLength)`)

roryabraham commented on src/libs/SuffixUkkonenTree/index.ts:
```diff
@@ -0,0 +1,183 @@
+function makeTree(numericSearchValues: Int8Array) {
+    const maxNodes = 2 * numericSearchValues.length;
+    // Allocate an ArrayBuffer to store all transitions (flat buffer), 4 bytes per transition (Uint32)
+    const transitionNodes = new Int32Array(maxNodes * ALPHABET_SIZE * 4);
```
Would it make sense to call this variable `internalNodes`? Is this variable representing all non-leaf nodes in the tree?

roryabraham commented on src/libs/SuffixUkkonenTree/index.ts:
```diff
@@ -0,0 +1,183 @@
+ */
+function makeTree(numericSearchValues: Int8Array) {
+    const maxNodes = 2 * numericSearchValues.length;
+    // Allocate an ArrayBuffer to store all transitions (flat buffer), 4 bytes per transition (Uint32)
```
Here's my attempt at explaining this variable better (but still trying to understanding all this)

```suggestion
    /* 
       This array represents all internal nodes in the suffix tree.
       When building this tree, we'll be given a character in the string, and we need to be able to lookup in constant time
       if there's any edge connected to a node starting with that character. For example, given a tree like this:

               root
              / | \
             a  b  c
 
       and the next character in our string is 'd', we need to be able do check if any of the edges from the root node
       start with the letter 'd', without looping through all the edges.

       To accomplish this, each node gets an array matching the alphabet size (x4 because it's 4 bytes per character).
       So you can imagine if our alphabet was just [a,b,c,d], then each node would get an array like [0,0,0,0].
       If we add an edge starting with 'a', then the root node would be [1,0,0,0]
       So given an arbitrary letter such as 'd', then we can take the position of that letter in its alphabet (position 3 in our example)
       and check whether that index in the array is 0 or 1. If it's a 1, then there's an edge starting with the letter 'd'.

       Note that for efficiency, all nodes are stored in a single flag array. That's how we end up with (maxNodes * alphabet_size * 4).
       In the example of a 4-character alphabet, we'd have an array like this:

          root       root.left     root.right              last possible node
        /     \       /     \       /     \                      /     \
       [0,0,0,0,      0,0,0,0,      0,0,0,0,  .................  0,0,0,0]
     */
```

Does that make sense? I know it's a lot of work, but that level of detail will be very valuable to make this code easier to understand imo.

Szymon20000 replied:
Sounds good. Normally we would store them in an array of maps. transition: Map<char, Node>[]. But to improve performance we first moved to Array of Arrays. transition: Number[][]. Then when we realised that ArrayBuffers are much faster (and they let as parallel the work as shared buffers can shared between different worklets runtimes) We switched to one dimension.

hannojg replied:
No, the sharing between worklet runtimes isn't soemthing thats happening here but something we are very eager to try soon!

hannojg replied:
Very nice explanation by the way Rory, adding that!

roryabraham commented on src/libs/SuffixUkkonenTree/index.ts:
```diff
@@ -0,0 +1,183 @@
+    const transitionNodes = new Int32Array(maxNodes * ALPHABET_SIZE * 4);
+    transitionNodes.fill(-1); // Initialize all transitions to -1 (no transition)
+
+    const leftEdges = new Int32Array(maxNodes * 4);
```
Why `maxNodes * 4`?

hannojg committed:
Update src/libs/SuffixUkkonenTree/utils.ts

Co-authored-by: Rory Abraham <47436092+roryabraham@users.noreply.github.com>

hannojg commented on src/libs/SuffixUkkonenTree/utils.ts:
```diff
@@ -0,0 +1,110 @@
+    },
+): {
+    numeric: Int8Array;
+    occurrenceToIndex: Int32Array;
```
> I can't imagine we'll be searching any strings longer than 65000 characters, so would that be slightly more efficient?

For my hightraffic account the search string has a length of 177k chars, i had the data from some customers and the string length was close to half a million.

roryabraham replied:
wow, 32 it is then

roryabraham replied:
Let's add a comment to explain the choice please

hannojg commented on src/libs/SuffixUkkonenTree/utils.ts:
```diff
@@ -0,0 +1,110 @@
+        clamp?: boolean;
+    },
+): {
+    numeric: Int8Array;
```
Yeah good point, i will double check all arrays and convert them to unsigned. For some we need them to be signed, for the transition nodes for example

hannojg commented on src/libs/SuffixUkkonenTree/utils.ts:
```diff
@@ -0,0 +1,110 @@
+
+/**
+ * Converts a string to an array of numbers representing the characters of the string.
+ * Every number in the array is in the range 0-ALPHABET_SIZE (0-28).
```
yeah, good point!

hannojg commented on src/libs/SuffixUkkonenTree/utils.ts:
```diff
@@ -0,0 +1,110 @@
+            occurrenceToIndex[offset.value] = index;
+            outArray[offset.value++] = SPECIAL_CHAR_CODE;
+            const asBase64Numeric = convertToBase26(charCode);
+            // eslint-disable-next-line @typescript-eslint/prefer-for-of
```
For browser environments for-indexed loops are known to be faster than for-of loops, quick check:

![Screenshot 2024-09-30 at 08 18 35](https://github.com/user-attachments/assets/47c3449a-6f7a-44ce-a611-859249444ff1).
As this loop will be potentially executed hundred thousands of time its important to have it as fast as possible.

I haven't explicitly checked this for hermes. As we run the code also on web, i think turning on that babel plugin does make sense!

Lets maybe do another PR first, where we enable the plugin?

hannojg replied:
moved the convo for discussing this change to [here](https://margelo.slack.com/archives/C035J5C9FAP/p1727770636895669?thread_ts=1727360885.983619&cid=C035J5C9FAP)

Szymon20000 commented on src/libs/SuffixUkkonenTree/index.ts:
```diff
@@ -0,0 +1,183 @@
+            }
+        }
+
+        dfs(0, 0);
```
Using iterative DFS is a better choice in compiled languages for instance C++. However, we noticed that V8 really liked splitting the code into smaller functions as then it understands it better and code is much faster. Also recursive approach turned out to be faster. Most likely it was caused by just in time compilation that V8 support. However, Hermes doesn't support JIT so it's possible that iterative approach is faster on it.

roryabraham replied:
whoa, that's surprising. I think we can leave it in its iterative form then

hannojg commented on src/CONST.ts:
```diff
@@ -1096,6 +1096,9 @@ const CONST = {
         RESIZE_DEBOUNCE_TIME: 100,
         UNREAD_UPDATE_DEBOUNCE_TIME: 300,
+        SEARCH_CONVERT_SEARCH_VALUES: 'search_convert_search_values',
+        SEARCH_MAKE_TREE: 'search_make_tree',
```
Yeah, i would like to collect as many as metrics in the beginning here as possible so we know where to optimize.

hannojg committed:
Update src/libs/FastSearch.ts

Co-authored-by: Rory Abraham <47436092+roryabraham@users.noreply.github.com>

hannojg committed:
Update src/libs/SuffixUkkonenTree/utils.ts

Co-authored-by: Rory Abraham <47436092+roryabraham@users.noreply.github.com>

hannojg committed:
Update src/libs/SuffixUkkonenTree/utils.ts

Co-authored-by: Rory Abraham <47436092+roryabraham@users.noreply.github.com>

hannojg committed:
Update src/libs/SuffixUkkonenTree/index.ts

Co-authored-by: Rory Abraham <47436092+roryabraham@users.noreply.github.com>

hannojg committed:
Update src/libs/SuffixUkkonenTree/index.ts

Co-authored-by: Rory Abraham <47436092+roryabraham@users.noreply.github.com>

Szymon20000 commented on src/libs/SuffixUkkonenTree/index.ts:
```diff
@@ -0,0 +1,183 @@
+ * The tree will be built using the Ukkonen's algorithm: https://www.cs.helsinki.fi/u/ukkonen/SuffixT1withFigs.pdf
+ */
+function makeTree(numericSearchValues: Int8Array) {
+    const maxNodes = 2 * numericSearchValues.length;
```
Every leaf represents a suffix. There can't be more than n suffixes. 
Every internal node has to have at least 2 children. So the total size of ukkonen tree is not bigger than 2n - 1.

roryabraham replied:
Thanks for explaining. Can you include this in a comment?

Szymon20000 commented on src/libs/SuffixUkkonenTree/index.ts:
```diff
@@ -0,0 +1,183 @@
+    const maxNodes = 2 * numericSearchValues.length;
+    // Allocate an ArrayBuffer to store all transitions (flat buffer), 4 bytes per transition (Uint32)
+    const transitionNodes = new Int32Array(maxNodes * ALPHABET_SIZE * 4);
+    transitionNodes.fill(-1); // Initialize all transitions to -1 (no transition)
```
We could use 0 but then we would need to move all the nodes as root is marked as 0.

ahmedGaber93 commented:
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

</details>

ahmedGaber93 commented:
This PR is a great performance improvement, but I think we missed some functionality that was in `filterOptions` (not sure if this intentionally or not) like [email without (.)](https://github.com/Expensify/App/blob/86360745431edb355c91e5594c741a54e5f88581/src/libs/OptionsListUtils.ts#L2459), [uique result](https://github.com/Expensify/App/blob/86360745431edb355c91e5594c741a54e5f88581/src/libs/OptionsListUtils.ts#L2472), [filters](https://github.com/Expensify/App/blob/86360745431edb355c91e5594c741a54e5f88581/src/libs/OptionsListUtils.ts#L2493), [order options](https://github.com/Expensify/App/blob/86360745431edb355c91e5594c741a54e5f88581/src/libs/OptionsListUtils.ts#L2516), searching by special chars and white spaces.

hannojg committed:
use Uint8Array for search numeric representation

hannojg committed:
fix utils

hannojg commented on src/libs/SuffixUkkonenTree/utils.ts:
```diff
@@ -0,0 +1,110 @@
+): {
+    numeric: Int8Array;
+    occurrenceToIndex: Int32Array;
+    offset: {value: number};
```
because offset is also passed around by reference, and we can't pass just number as reference

hannojg committed:
add comment

hannojg committed:
Update src/libs/SuffixUkkonenTree/index.ts

Co-authored-by: Rory Abraham <47436092+roryabraham@users.noreply.github.com>

hannojg committed:
implementation using none negative values

implementation using none negative values

Co-Authored-By: Szymon Kapała <szymon20000@gmail.com>

hannojg committed:
adjust array sizes to be correct

hannojg committed:
fix: can't find by first letter

hannojg committed:
remove unnecessary initializations

hannojg committed:
fix: can't find by first letter was only a problem in the unit test

hannojg committed:
cleanup

hannojg commented on src/libs/SuffixUkkonenTree/utils.ts:
```diff
@@ -0,0 +1,110 @@
+    occurrenceToIndex: Int32Array;
+    offset: {value: number};
+} {
+    const outArray = options?.out?.outArray ?? new Int8Array(input.length * 8); // We assume that the number output array will fit in 8 times the input length (for letters a-z only 1 number is needed, for any other unicode more numbers are needed)
```
We are very positive that the assumption won't be wrong. This gives us 8 extra letters per char for encoding a char to base26. The char code of the letter we'd encode would need to be greater than `8031810176` (26^(8 - special letter char)). [The highest code point is](https://www.compart.com/en/unicode/block) 0x0010FFFD = 1114109, which in base26 would only take 5 letters.
So actually 8 is already too much - will update it to 6 to be sure and update the comment

hannojg committed:
explain in comments

hannojg commented:
@roryabraham re: https://github.com/Expensify/App/pull/48652#discussion_r1781745534

we made the change to convert the transition nodes to an unsigned array. We were thus able to skip filling the array with `-1`, which for larger strings, can indeed be a significant performance gain.
The code however became a bit harder to understand now, as everything needed to be offset by 1.

Main change has been added here: https://github.com/Expensify/App/pull/48652/commits/fdac05fb311c3538f2929f35b1857e1317666866

hannojg committed:
+1 not +2

hannojg commented:
@ahmedGaber93 

- searching for [email without (.)](https://github.com/Expensify/App/blob/86360745431edb355c91e5594c741a54e5f88581/src/libs/OptionsListUtils.ts#L2459) is still supported, you can confirm by testing
- [uique result](https://github.com/Expensify/App/blob/86360745431edb355c91e5594c741a54e5f88581/src/libs/OptionsListUtils.ts#L2472) it would actually take longer to filter the results for unique then to just insert a duplicate in the tree. I added kind of an optimisation here where i check whether the display name is different from the login (because the display name seems to default to the email very often): https://github.com/margelo/expensify-app-fork/blob/29a193442e2f2913f5333c6b8064b4dc44c68b9f/src/pages/ChatFinderPage/index.tsx#L100
- the other stuff we are currently missing yet, will update the PR accordingly in a moment!

hannojg commented:
@ahmedGaber93 for the lag, are you able to record a performance profile for me in chrome and send it to me please?

hannojg committed:
order options

hannojg committed:
skip white spaces

hannojg commented:
@ahmedGaber93 the parsing of phone number was originally added to fix this case:

https://github.com/Expensify/App/issues/28492

- enter a valid US mobile number
- expect it to show in results
- make that valid US mobile number invalid
- expect: The search result will only be displayed if it exactly matches the search number.
- actual: The search result continues to be displayed even though it doesn't match the search number.

this case is already fixed automatically when using the suffix tree implementation!

hannojg committed:
don't return PD, concat with recentReports

hannojg commented:
@ahmedGaber93 for [filters](https://github.com/Expensify/App/blob/86360745431edb355c91e5594c741a54e5f88581/src/libs/OptionsListUtils.ts#L2493) i think thats not needed either. In the screenshot you can see that the logic basically sorts/orders the data twice. We can optimize that to only happen once.

In this commit i added the logic which concats the personalDetails results to the recent reports:
41207c61f922195dff6f70e3d77ce668540d415d

i'll be honest, i don't understand why we do that. The search results look the same to me before and after the commit. Any ideas why this is needed?

![Screenshot 2024-10-01 at 16 24 49](https://github.com/user-attachments/assets/809e6117-3e56-48ee-941b-c73e542c1efe)

hannojg commented:
_Note:_ Currently i moved the logic for filtering only into the `ChatFinderPage`. In the next Pr, where i want to create the fast search instance once and store it globally, i will look into how we can bring back the search filtering optimization to the `OptionListUtils` and apply it to all searches.

ahmedGaber93 commented:
[Google Chrome performance profile](https://github.com/user-attachments/files/17213521/Trace-20241001T081231.json.zip) the lag is from second 8 to 12 

> searching for [email without (.)](https://github.com/Expensify/App/blob/86360745431edb355c91e5594c741a54e5f88581/src/libs/OptionsListUtils.ts#L2459) is still supported, you can confirm by testing

It not works for me, result from `a.gaber` should be same as `agaber` without (.), I think we need to add `option.login?.replace(CONST.EMAIL_SEARCH_REGEX, '')` to `toSearchableString` in both [recentReports](https://github.com/margelo/expensify-app-fork/blob/29a193442e2f2913f5333c6b8064b4dc44c68b9f/src/pages/ChatFinderPage/index.tsx#L106) and [personalDetails](https://github.com/margelo/expensify-app-fork/blob/29a193442e2f2913f5333c6b8064b4dc44c68b9f/src/pages/ChatFinderPage/index.tsx#L100)

> In this commit i added the logic which concats the personalDetails results to the recent reports ... The search results look the same to me before and after the commit. Any ideas why this is needed?

I think this related to the `orderOptions` function, instead of displaying all ordered reports then all ordered personal details, we combine them into one array then ordering the combined arrays together.

And for [uniqFast](https://github.com/Expensify/App/blob/86360745431edb355c91e5594c741a54e5f88581/src/libs/OptionsListUtils.ts#L2472) I think I misunderstand it, It is not used for unique result, it used for [ranking](https://github.com/Expensify/App/blob/86360745431edb355c91e5594c741a54e5f88581/src/libs/filterArrayByMatch.ts#L93), and in this PR I don't think it will be affected if the searchable text is `txt1,txt2` or `txt1,txt1,txt2` when we search for `txt1`.

hannojg commented:
Thanks for sharing the profile! Could you try reproducing that on main as well please?

I had a look at it and the lag doesn't come from this PR. The lag is happening in `OptionListContextProvider`:

![Screenshot 2024-10-02 at 09 03 25](https://github.com/user-attachments/assets/aa9de788-5ecd-4d2b-ae1b-bb22c2f96d20)

We actually have an issue open for this, and your profile was very important and insightful, thank you so much:

- https://github.com/Expensify/App/issues/46593

ahmedGaber93 commented:
Yeah, that's true. It is a different issue, and I can reproduce it on main.

hannojg commented:
Ah yes, sorry, there were so many items to discuss that this one slipped me. Will add proper handling in this PR! (the PR shouldn't break existing functionality)

roryabraham commented:
Just as a heads up, the ChatFinderPage will soon be deprecated in favor of the SearchRouter component. So we should target this fix towards the new component instead.

hannojg commented:
I was wondering already why there were never any merge conflicts 💀 thanks for the heads up, will adjust accordingly!

hannojg committed:
Update src/libs/SuffixUkkonenTree/index.ts

Co-authored-by: Rory Abraham <47436092+roryabraham@users.noreply.github.com>

hannojg committed:
Merge branch 'main' of github.com:Expensify/App into perf/search-suffix-ukkonen-tree

hannojg committed:
Merge branch 'perf/search-suffix-ukkonen-tree' of github.com:margelo/expensify-app-fork into perf/search-suffix-ukkonen-tree

hannojg committed:
disable .at() rule

hannojg committed:
rename parameter

hannojg committed:
step A: bring changes over to SearchRouter
Todo: unify changes

hannojg committed:
reorder

hannojg committed:
fix dot use case

hannojg commented:
@ahmedGaber93 I fixed the dot use-case with emails - this is good for another review!

(I am currently thinking about the next PR, and how to unify the logic between SearchRouter, OptionUtilsList, and FastSearch... 

- right now we always recreate the fastSearch/suffixtree from scratch when the personal details or reports are changing. We can make this way more efficient by updating the existing instance (The `OptionListContextProvider` feels like a good place to do that, as it already has a logic for "create once, otherwise update")
- right now the filtering of the data (FastSearch) is happening outside of OptionUtilsList - i think it would be good to bring it back there (however, how do we do that if we only really want to keep one FastSearch instance per search-case... - will it be enough to have one FastSearch instance in `OptionListContextProvider` and reuse that across search use cases)

ahmedGaber93 reviewed (approved):
LGTM! The only different from `filterOptions` can I notice is searching with special characters (e.g. result is the same for `ahmed@` and `ahmed`) but this added intentionally [here](https://github.com/Expensify/App/pull/48652/files#diff-a919e8da171bd4fc998da5203188843b1e0587eb3fb2c809e9f235ce54cdf8cbR19) So I leave the decision to the internal developer.

hannojg commented on src/components/Search/SearchRouter/SearchRouter.tsx:
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable rulesdir/prefer-at */
```
ah sorry, i think we should migrate components when we work on them right? Will update

hannojg committed:
Merge branch 'main' of github.com:Expensify/App into perf/search-suffix-ukkonen-tree

hannojg committed:
remove eslint disabling

marcaaron commented:
I'm back from OOO now too so happy to jump in here.

marcaaron commented on src/components/Search/SearchRouter/SearchRouter.tsx:
```diff
@@ -56,6 +58,49 @@ function SearchRouter() {
+                data: searchOptions.personalDetails,
+                toSearchableString: (option) => {
+                    const displayName = option.participantsList?.[0]?.displayName ?? '';
+                    return [option.login ?? '', option.login !== displayName ? displayName : ''].join();
```
Does it matter that we will have a string with like `,something` if the login is an empty string or just `,` if both are empty? Not sure if it can happen in practice, but maybe we should not include any empty strings at all.

hannojg replied:
empty strings won't be searchable. No characters would be added for that to the search tree, thus they'd never be retrievable

marcaaron commented on src/libs/FastSearch.ts:
```diff
@@ -0,0 +1,140 @@
+const charSetToSkip = new Set(['@', '.', '#', '$', '%', '&', '*', '+', '-', '/', ':', ';', '<', '=', '>', '?', '_', '~', '!', ' ']);
+
+/**
+ * Creates a new "FastSearch" instance. "FastSearch" uses a suffix tree to search for (sub-)strings in a list of strings.
```
```suggestion
 * Creates a new "FastSearch" instance. "FastSearch" uses a suffix tree to search for substrings in a list of strings.
```

?

marcaaron commented on src/libs/FastSearch.ts:
```diff
@@ -0,0 +1,140 @@
+ */
+function createFastSearch<T>(dataSets: Array<SearchableData<T>>) {
+    Timing.start(CONST.TIMING.SEARCH_CONVERT_SEARCH_VALUES);
+    const maxNumericListSize = 400_000;
```
Could we exceed this? Or is that pretty unlikely? What happens if we do?

hannojg replied:
Good question. I made this number to be based on the most extreme case i have seen so far (the search string generated for ten thousands of reports and personal details was below 400.000, don't remember the exact number anymore).
In case the search string is longer, we'd loose search results. I think no error would be thrown, but if the user you are searching for is on position 400.010 you wouldn't be able to find that person.

Let me brain storm quickly with Szymon if we can do better!

hannojg replied:
We think it would be best to implement a resize mechanism. If we hit the limit, we'd resize the arrays to be bigger. However, that comes with a performance overhead. So we should pick a default size that makes sense for most users (400_00 might actually be a bit too much for the average case, but for some extreme cases it might be too little).
However, as this isn't directly breaking anything, would you be fine with us tweaking this in a follow up PR? I think the Pr is already quite huge and hard to understand

hannojg replied:
(Another solution for performance improvements on this could also be to store the array's arraybuffer in mmkv and rehydrate from there [mmkv has first class support for ArrayBuffers], btu again, it would be better to experiment with that in a follow up PR - i can open a follow up issue ticket, i think that would be best?)

danieldoglas replied:
I think it's fine to keep it as 400K for now, but have a follow-up issue to make it resizable.

marcaaron replied:
Ok, thanks! I think this is something we can wait and see about - my initial thinking is that we still don't have any pagination for either personal details or reports. And we have also spent a lot of time tweaking things for some very high capacity usage.

Though, it was hard for me to tell how many personal details or reports we are talking about here when we'd be hitting that limit. If it's over 10k reports then nobody actually has that AFAIK.

marcaaron commented on src/libs/FastSearch.ts:
```diff
@@ -0,0 +1,140 @@
+            const dataSetIndex = listOffsets.findIndex((listOffset) => occurrenceIndex < listOffset);
+
+            if (dataSetIndex === -1) {
+                throw new Error('Programmatic error, this should never ever happen');
```
If it should never ever happen then why is the code necessary ? :trollface:

Maybe we say something less generic? Like, `Oh no. Ask Hanno how this code works.`

Just kidding, but seriously maybe "Search failed due to range error" or something.

danieldoglas replied:
I agree with more specific errors. Considering right now we have 2 with the same text, it would be hard to get to know where it failed.

hannojg replied:
Yes, let me update this. Actually this might be removable completely. Its a reminiscent of when the dataSetIndex could be `undefined`, which we knew could never happen and we only added that statement to satisfy typescript.

However, i might keep the checks in place because .... you never know, right

marcaaron commented on tests/unit/FastSearchTest.ts:
```diff
@@ -0,0 +1,118 @@
+        expect(search('test.user')).toEqual([['test.user@example.com']]);
+        expect(search('examplecom')).toEqual([['test.user@example.com']]);
+    });
+});
```
These tests are good. Should we test some common formats of reports and personal details? Feels like there are a few recurring patterns that could be easily added here.

marcaaron replied:
Ok. Sounds like the benefit of doing it now is that the tests would help catch potential issues?  I don't think it matters as much as long as all the mainline cases for searching are working (which they seem to be). Let's maybe just plan on doing that follow up plan with some urgency.

marcaaron reviewed (commented):
Looks pretty thorough and well documented. I am not too familiar with this algorithm so trust that it does what it should. I wouldn't mind seeing a few more tests - maybe using some mock `personalDetails` and `recentReports`.

danieldoglas reviewed (commented):
LGTM overall, just some small comments.

danieldoglas commented:
bump @hannojg, do you think you can address the comments until EOW?

hannojg commented:
@danieldoglas absolutely, will update the PR tomorrow (CET)!

hannojg committed:
Merge branch 'main' of github.com:margelo/expensify-app-fork into perf/search-suffix-ukkonen-tree

hannojg committed:
Update src/libs/FastSearch.ts

Co-authored-by: Marc Glasser <marc.aaron.glasser@gmail.com>

hannojg committed:
improve error messages

hannojg committed:
Update src/libs/SuffixUkkonenTree/utils.ts

Co-authored-by: Daniel Silva <danieldoglas@users.noreply.github.com>

hannojg commented on src/components/Search/SearchRouter/SearchRouter.tsx:
```diff
@@ -56,6 +58,49 @@ function SearchRouter() {
+     * Builds a suffix tree and returns a function to search in it.
+     */
+    const findInSearchTree = useMemo(() => {
+        const fastSearch = FastSearch.createFastSearch([
+            {
+                data: searchOptions.personalDetails,
+                toSearchableString: (option) => {
+                    const displayName = option.participantsList?.[0]?.displayName ?? '';
+                    return [option.login ?? '', option.login !== displayName ? displayName : ''].join();
+                },
+            },
+            {
+                data: searchOptions.recentReports,
+                toSearchableString: (option) => {
+                    const searchStringForTree = [option.text ?? '', option.login ?? ''];
+
+                    if (option.isThread) {
+                        if (option.alternateText) {
+                            searchStringForTree.push(option.alternateText);
+                        }
+                    } else if (!!option.isChatRoom || !!option.isPolicyExpenseChat) {
+                        if (option.subtitle) {
+                            searchStringForTree.push(option.subtitle);
+                        }
+                    }
+
+                    return searchStringForTree.join();
+                },
+            },
```
100%, this is actually what i want to do in the next PR!

I wouldn't want to change that now, because:

- i am addressing this in the next PR
- the `ChatFinderPage` is going to be deprecated very soon, so then this code would only exist in `SearchRouter`

is that fine with you? 😊

danieldoglas replied:
Seems fine to me!

hannojg committed:
fix variable name usage

danieldoglas reviewed (dismissed):
LGTM

danieldoglas commented:
@ahmedGaber93 can you please test this again to confirm everything is working? And then we can merge it by the end of the day!

ahmedGaber93 reviewed (approved):
LGTM! tested well

danieldoglas commented:
@marcaaron all yours to a last look and merge!

marcaaron reviewed (dismissed):
LGTM 👍

marcaaron commented:
Re-running the perf test - I would first assume it failed for some other reason.

marcaaron commented:
Hmm still failing.

hannojg commented:
Sorry for not posting it earlier, the discussion has been here about the failed performance test. I think the conclusion was to do nothing here in this PR and let Fabio handle it:

- https://margelo.slack.com/archives/C01GTK53T8Q/p1727461806388989

danieldoglas commented:
@hannojg can you please solve the conflicts here?

@marcaaron once the conflicts are solved, do you think we can merge this? I agree that if the performance issue is with the change to `useOnyx`, we should treat it separately from this PR

marcaaron commented:
Yes, but I am waiting for confirmation about how the performance tests work? I am unsure if merging this means that all App PRs will start to fail. Asked about it [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1729106669701779?thread_ts=1727461806.388989&cid=C01GTK53T8Q).

hannojg committed:
Merge branch 'main' of github.com:Expensify/App into perf/search-suffix-ukkonen-tree

hannojg committed:
fix duplicate imports after merge

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

IuliiaHerets commented:
This PR is failing for Android because of issue #51123
